### ISSUE TITLE
when the client charset is UTF8, perl native strings should be automatically upgraded to utf8

### ISSUE THREAD

**Issue by damil** [perl5-dbi/DBD-Oracle#161]
(cross-posted from https://stackoverflow.com/questions/75245442/how-do-i-handle-unicode-with-dbdoracle)

The DBI documentation says :

> Perl supports two kinds of strings: Unicode (utf8 internally) and non-Unicode (defaults to iso-8859-1 if forced to assume an encoding). Drivers should accept both kinds of strings and, if required, convert them to the character set of the database being used. Similarly, when fetching from the database character data that isn't iso-8859-1 the driver should convert it into utf8.

DBD::SQLite and DBD::Pg indeed do upgrade Perl native strings to utf8 strings before sending them to the database, for example in INSERTs or UPDATEs. DBD::Oracle (with NLS_LANG set to UTF8) does not. So with Oracle it seems to be the responsability of the application to distinguish between native and utf8 strings and do the proper encoding.

Is this singularity an omission, or is it by design ?

The stackoverflow link above contains a test program to illustrate the issue.

**Comment by damil** [comment]
Some additional information : 

1. the fact that DBD::Oracle seems to ignore Perl's internal utf8 flag was already mentioned ... in 2007 (https://www.perlmonks.org/?node_id=649489), with no satisfactory answer
2. Looking at the implementation of DBD::Pg, this driver makes  a deliberate effort to "do the right thing" with Perl strings : the driver has a routine 'pg_rightgraded_sv' (https://metacpan.org/release/TURNSTEP/DBD-Pg-3.16.0/source/dbdimp.c#L2974) which decides when to do utf8::upgrade or utf8::downgrade. That routine is called from prepare(), bind_param() and many other functions. The test file https://metacpan.org/release/TURNSTEP/DBD-Pg-3.16.0/source/t/30unicode.t performs extensive tests on that feature.

So it seems to me that DBD::Oracle should do something similar to make the client's life easier. Otherwise, it means that higher-level application layers must do the appropriate string transformations ... not very convenient. 

If no solution comes from DBD::Oracle, my backup plan will be to exploit DBI's callback mechanism to insert my own string upgrading routines before calls to prepare(), execute(), bind_param(), etc. Some preliminary tests demonstrate that this will solve the problem, but it would be sooooo much better if DBD::Oracle did the job !

**Comment by djzort** [comment]
This is unlikely to be by design, almost certainly its an omission.

Perhaps someone can use the DBD::Pg code as inspiration for a pull request?

**Comment by damil** [comment]
@djzort : thanks for the comment.
With pleasure I would have proposed a pull request ... but looking at the internals of DBD::Oracle, it looks like a very complicated and risky task. So it will have to be somebody else who is at ease with that code !
Meanwhile, I have written a workaround using DBI callbacks -- see https://metacpan.org/pod/DBIx::Oracle::UpgradeUtf8.
This is not the ideal solution, but at least it is no longer a blocking issue for me -- I can wait for a future release of DBD::Oracle

**Comment by whindsx** [comment]
If somebody is considering writing a pull request, they will probably want to use the [cand-v1.90](https://github.com/perl5-dbi/DBD-Oracle/tree/cand-v1.90) branch since it contains significant changes to dbdimp.c....

**Comment by djzort** [comment]
Its very close to release, i am just hoping someone can fix up https://github.com/perl5-dbi/DBD-Oracle/issues/59

**Comment by damil** [comment]
Is there any company offering paid support for DBD::Oracle ? Maybe I could convince my employer to find a budget for sponsoring the job.

**Comment by damil** [comment]
I studied the situation deeper and came to the conclusion that even if DBD::Oracle did automatically upgrade native strings, as DBD:SQLite or DBD::Pg do, it would not completely resolve the problem. 

Perl utf8::upgrade assumes an iso-8859-1 charset; so if this is performed automatically, it is not convenient when the native charset is something different, like for example Windows-1252. So I wrote a new CPAN module that lets client explicitly specify what is the charset to decode from : https://metacpan.org/pod/DBIx::AutoUpgrade::NativeStrings . 

Since I no longer consider that DBD::Oracle should absolutely do the upgrade, I'll close the issue.
